### ISSUE TITLE
Incorrect labelling of unreachable code

### ISSUE THREAD

**Issue by lazarillo** [microsoft/pyright#10047]
## Environment data

-   Pylance version: 2025.3.1
-   OS and version: Mac OS X 15.3.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.11

## Code Snippet

```python
import google.auth
from google.auth.exceptions import (
    DefaultCredentialsError,
    UserAccessTokenError,
    RefreshError,
)

def check_gcloud_auth() -> bool:
    """
    Checks if gcloud authentication is properly configured.

    Returns:
        bool: True if authentication is likely configured, False otherwise.
    """
    try:
        credentials, _project_id = (
            google.auth.default()
        )  # Attempt to get default credentials
        if credentials and credentials.valid:
            return True
        else:
            return False
    except (DefaultCredentialsError, UserAccessTokenError, RefreshError):
        return False  # Default credentials not found
    return False
```

## Expected behavior

The code should not have any errors or issues.

## Actual behavior

The final `return False` is darkened and `pylance` is flagging it as unreachable (`Code is unreachable - Pylance`).

The final `return False` can, in fact, be reached any time an exception is experienced that is *not* one of these three exceptions.

**Comment by erictraut** [comment]
Upon further inspection, pyright is working correctly here.

> The final return False can, in fact, be reached any time an exception is experienced that is not one of these three exceptions.

That's an incorrect assumption. If an exception is raised that doesn't match any of these three exceptions, control will not return to this function. The raised exception will propagate up the stack, and the final `return` statement will never be reached.

You can see this by running the following sample:

```python
def func():
    try:
        x = 1 / 0
        return
    except ValueError:
        return
    print("Never gets here")

func()
```

If you want to catch other exception types, you'd need to add an explicit `except:` clause or a `finally` clause.
